### ISSUE TITLE
Pug shares the scope between multi-level templates in a strange way

### ISSUE THREAD

**Issue by NoNameProvided** [pugjs/pug#2703]
I don't know if this is the expected behaviour or not, but I would like to start a discussion about the scope of the `extend` & `include` commands are share.

Now the current behaviour is that a pug file which extends an another shares it's scope in a strange way with it's included templates. I find this behaviour odd, as it's restrict your structure to be one level deep maximum. 

Here is an example, having the following three files in a directory:

```pug
//- base.pug
doctype html
html
  head
    block options
    - var title = (options && options.title) ? options.title : "Default title"
    //- ...
    link(rel='stylesheet' href='/bundle.css')
    //- ...
    base(href='/')

    title #{title}

  body
    block content
    script(src='/shared-bundle.js')
    block scripts

```

```pug
//- component.pug
block scripts
  script(src='/js/footer.component.js')

div.c-component
  p Content of the component.
```

```pug
//- page.pug
extends ./base.pug

block options
  - var options = { title: 'Pug Test' }

block scripts
  script(src='/js/test.page.js')

block content
  .p-test
    p Content of the page.
    include ./component.pug
```

This will generate the following html when running the `pug --pretty --out dist/ page.pug` command with the _pug-cli_.

```html
<!DOCTYPE html>
<html>
  <head>
    <link rel="stylesheet" href="/bundle.css">
    <base href="/">
    <title>Pug Test</title>
  </head>
  <body>
    <div class="p-test">
      <p>Content of the page.</p>
      <script src="/js/footer.component.js"></script>
      <div class="c-component">
        <p>Content of the component.</p>
      </div>
    </div>
    <script src="/shared-bundle.js"></script>
    <script src="/js/footer.component.js"></script>
  </body>
</html>
```

As you can see the `footer.component.js` file which was defined in the `component.pug` is included twice. Once (in my opinion) in it's right place at the end of the html file inside the `base.pug` scripts section, but its included in an another place before the component.

So the following snippet in the `component.pug` template cause two strange side effect:

```pug
block scripts
  script(src='/js/footer.component.js')
```

First the `footer.component.js` file is included twice, second the `/js/test.page.js` file defined in the page template is not included.

If I remove the `scripts` block from the component.pug and simply add the script tag, then the `/js/test.page.js` file gets included properly at it's place.

This behaviour is clearly odd, and I think it's a bug. If not then the reason for this should be explained in the docs.

**Comment by ForbesLindesay** [comment]
> First the footer.component.js file is included twice

This looks like it's mostly the same issue as #2687. Although the behaviour is certainly unintuitive, I'm not sure it's a bug as such.  The issue is that when we see the block in the include, we're not sure if you're attempting to override a block in a parent, or declare a block, for child templates to replace.

> second the /js/test.page.js file defined in the page template is not included.

```pug
block scripts
  script(src='/js/footer.component.js')
```

means "replace the block, scripts, with this content".  You should use either `append` or `prepend` (depending on what resulting order you want:

```pug
append scripts
  script(src='/js/footer.component.js')
```

This part of the behaviour is not a bug.

> The expected behaviour would be to be able to use blocks in in every child in the hierarchy

You can use blocks in any level of a hierachy that's a series of `extends`,  i.e.

```pug
// home.pug

extends ./page-layout.pug

block content
  h1 My Home Page
```

```pug
// page-layout.pug

extends ./layout.pug

block body
  block content
  script(src="./page-script.js")
```


```pug
// layout.pug

html
  body
    block body
```

**Comment by NoNameProvided** [comment]
> The issue is that when we see the block in the include, we're not sure if you're attempting to override a block in a parent, or declare a block, for child templates to replace.

It makes sense, but then I would think that `prepend` / `append` should work, but the result is the same. The script got included twice for the following pug: 

```
append scripts
  script(src='/js/header.component.js')
```

Although writing out `append` (in my opinion) should indicate I want to attach to a block higher in a parent. Wouldn't it make sense?

> The issue is that when we see the block in the include, we're not sure if you're attempting to override a block in a parent, or declare a block, for child templates to replace.

Wouldn't introducing `replace` solve this problem? This way if the compiler finds a `replace`, `append` or `prepend` it knows it always want to write into the  parent's `block` and when find a `block` it knows a new block is declared.

**Comment by ForbesLindesay** [comment]
Yes, adding `replace` would resolve this issue.  It's a big part of what we want to do for v3.0.0, but it will involve a number of breaking changes that I don't think we want to make as part of this release.

**Comment by NoNameProvided** [comment]
Yeah, that is understandable. Is there any schedule for v3?

**Comment by ForbesLindesay** [comment]
Not really, but unless enough people become [backers/sponsors](https://opencollective.com/pug) for me to work on this full time, I would imagine a year or two.  Only just started attempting to raise funds though, so you never know.

**Comment by raduluchian** [comment]
I've been struggling to figure out why the Dev Tools console was throwing errors attributed to duplicated IDs in my code. At first I thought I was doing duplicated imports in my Pug templates but as it turns out it's actually the very same issue described by @NoNameProvided and I am actually using the `append` method.

Long story short: I have a master template that has a `block aside` defined and then a set of different product templates, each comprised out of different sub-pages. It's these sub-pages that are trying to append content to the that master `aside` ending up in duplicated imports. 

Obviously the very same thing was happening for `block scripts` and `block styles` but I didn't realize it because no error were thrown for loading the same file twice.

**Comment by muratsu** [comment]
Is there a workaround for this issue until v3 gets released? We have a structure similar to @NoNameProvided & @raduluchian and not sure what to do. ☹️

**Comment by Robula** [comment]
The only way I believe we can work around this is to avoid nested block append inside included blocks.. 😕

**Comment by bbayo4** [comment]
It's end of 2018. Any fix for this?

**Comment by hiiphop** [comment]
Hello, I have had same problem. Did anyone fix it ?

**Comment by brolnickij** [comment]
It's end of 2019. Any fix for this? :(

**Comment by winterNebs** [comment]
Just ran into this as well. 

For my use case I can get around it by splitting the include into separate files and including them separately, but that's a hacky solution that won't always work.

**Comment by omar84** [comment]
mid 2020, corona will have a vaccine and this is still not fixed!

**Comment by phoenixweiss** [comment]
Hello from mid 2021!

**Comment by trasherdk** [comment]
[Maybe this could help](https://github.com/pugjs/pug/issues/2703#issuecomment-279417806)

**Comment by NoNameProvided** [comment]
Hi all!

> mid 2020, corona will have a vaccine and this is still not fixed!

I am sure if anyone picks this up proposes a way to fix this, verify it with Forbes and then implement the fix then it will get merged. Crying in an open-source repo about someone not giving their free-time to you is counter-productive. Please don't increase the noise, act if you need it or wait otherwise. 

Thanks.
